packages/hub/services/crank-nonce-lock.ts

```diff
@@ -0,0 +1,65 @@
+import { inject } from '@cardstack/di';
+import { Prisma } from '@prisma/client';
+import BN from 'bn.js';
+import config from 'config';
+import { addSeconds } from 'date-fns';
+import { Wallet } from 'ethers';
+import { nowUtc } from '../utils/dates';
+
+export default class CrankNonceLock {
+  prismaManager = inject('prisma-manager', { as: 'prismaManager' });
+  ethersProvider = inject('ethers-provider', { as: 'ethersProvider' });
+  CRANK_LOCK_KEY = 1;
+
+  private async getNonce(chainId: number, dbTransaction: Prisma.TransactionClient) {
+    let crankNonce = await dbTransaction.crankNonce.findUnique({
+      where: { chainId: chainId },
+    });
+    let one = new BN(1);
+    let nextNonce = crankNonce?.nonce ? new BN(crankNonce.nonce.toString()).add(one) : new BN(0);
+    let nonceTTL = Number(config.get('nonceTTL'));
+    let now = nowUtc();
+    if (nextNonce.cmp(one) <= 0 || (crankNonce && addSeconds(crankNonce.updatedAt, nonceTTL) <= now)) {
+      let crank = new Wallet(config.get('hubPrivateKey'));
+      let provider = this.ethersProvider.getInstance(chainId);
+      nextNonce = new BN(await provider.getTransactionCount(crank.address, 'latest'));
+    }
+
+    await dbTransaction.crankNonce.upsert({
+      where: { chainId: chainId },
+      create: {
+        chainId,
+        nonce: BigInt(nextNonce.toString()),
+      },
+      update: {
+        nonce: BigInt(nextNonce.toString()),
+      },
+    });
+
+    return nextNonce;
+  }
+
+  async withNonce(chainId: number, cb: (nonce: BN) => Promise<any>) {
+    let prisma = await this.prismaManager.getClient();
+    await prisma.$executeRaw`SELECT pg_advisory_lock(${this.CRANK_LOCK_KEY})`;
+    //Make database transaction here
+    //because if the execution failed the nonce should be rollback
+    try {
+      return await prisma.$transaction(
+        async (tx) => {
+          let nonce = await this.getNonce(chainId, tx);
+          return await cb(nonce);
+        },
+        { maxWait: 3000, timeout: 10000 }
```

Comment:
Can you explain why this is needed and on what these numbers are based on?

Reply:
- `maxWait`: The maximum amount of time the Prisma Client will wait to acquire a transaction from the database. The default value is 2 seconds.
- `timeout:` The maximum amount of time the interactive transaction can run before being canceled and roll back. The default value is 5 seconds.

I increased the value of `maxWait` and `timeout` because we need to wait for the execution transaction to be published, and that process may take more than 5 seconds.

Reply:
Does this mean that transaction has to be published in under 10 seconds, otherwise the transaction will fail? Is it normal this should happen under 10s, even on mainnet?

Reply:
No, it means the database transaction will be acquired by this process for 10 seconds. If publishing the transaction to the blockchain takes more than 10 seconds, the database transaction will be released/rollback.

We only wait until the transaction is published, not mine, so it must be lower than 10 seconds.

Reply:
>  If publishing the transaction to the blockchain takes more than 10 seconds, the database transaction will be released/rollback.

This is the part I am worried about. What makes us think it shouldn't take more than 10 seconds to publish the transaction? What about mainnet? Or when the network is really congested? 

p.s. whatever conclusion comes from this debate please add a comment in the code explaining it so the next person can see why/how we got these numbers.

Reply:
the act of sending a txn is a normal web request. these generally have responses in less than 1 second. A server that takes 10 seconds to respond is incredibly degraded.

Reply:
keep in mind that sending a txn is not the same as waiting for a txn to be mined (aka consensus). you can send a txn to any RPC node--so network congestion in the normal sense doesn't have any impact here since we are talking about a  decentralized system. there is not like a single RPC node that is clogged up when the network traffic is high. Albiet, we might be unlucky enough to be talking to an RPC node that is over burdended by a lot of clients--which is something that can happen even when network congestion for the protocol is low.

Reply:
👍 Thanks, I'm fine with that then.

Reply:
Added comment to the code